src/main/java/com/hermesworld/ais/galapagos/changes/impl/ChangeBase.java

```diff
@@ -338,6 +346,79 @@ protected boolean isEqualTo(ChangeBase other) {
 
 }
 
+@JsonSerialize
+final class TopicProducerAddChange extends ChangeBase {
+
+    private final String topicName;
+
+    private final List<String> topicProducerIds;
+
+    public TopicProducerAddChange(String topicName, List<String> topicProducerIds) {
+        super(ChangeType.TOPIC_PRODUCER_APPLICATION_ADDED);
+        this.topicName = topicName;
+        this.topicProducerIds = topicProducerIds;
+
+    }
+
+    public String getTopicName() {
+        return topicName;
+    }
+
+    public List<String> getTopicProducerIds() {
+        return topicProducerIds;
+    }
+
+    @Override
+    protected boolean isEqualTo(ChangeBase other) {
+        return Objects.equals(topicName, ((TopicProducerAddChange) other).topicName);
+    }
+
+    @Override
+    public CompletableFuture<?> applyTo(ApplyChangeContext context) {
+        return context.getTopicService().addTopicProducers(context.getTargetEnvironmentId(), topicName,
+                topicProducerIds);
+    }
+
+}
+
+@JsonSerialize
+final class TopicProducerRemoveChange extends ChangeBase {
+
+    private final String topicName;
+
+    private final List<String> topicProducerDeleteId;
+
+    public TopicProducerRemoveChange(String topicName, List<String> topicProducerDeleteId) {
+        super(ChangeType.TOPIC_PRODUCER_APPLICATION_REMOVED);
+        this.topicName = topicName;
+        this.topicProducerDeleteId = topicProducerDeleteId;
+
+    }
+
+    public String getTopicName() {
+        return topicName;
+    }
+
+    public List<String> getTopicProducerIds() {
+        return topicProducerDeleteId;
+    }
+
+    @Override
+    protected boolean isEqualTo(ChangeBase other) {
+        return Objects.equals(topicName, ((TopicProducerRemoveChange) other).topicName);
+    }
+
+    @Override
+    public CompletableFuture<?> applyTo(ApplyChangeContext context) {
+
+        topicProducerDeleteId.forEach(deleteId -> context.getTopicService()
+                .removeProducerFromTopic(context.getTargetEnvironmentId(), topicName, deleteId));
```

Comment:
Very dangerous - Future return value is ignored. You **must** return the futures somehow.
When you change implementation (as suggested above) to ONE removed producer only, this gets much simpler here.